lib/ecto/query.ex

```diff
@@ -757,12 +757,40 @@ defmodule Ecto.Query do
       |> select([g, gs], {g.name, gs.sold_on})
 
   """
-  defmacro join(query, qual, binding \\ [], expr, on \\ nil) do
+  defmacro join(query, qual, binding \\ [], expr, opts \\ []) do
+    opts =
+      case parse_join_opts(opts) do
+        {:opts, opts} ->
+          opts
+        {:expr, on_expr} ->
+          IO.warn "Passing raw `on` expression as the last argument to Ecto.Query.join/5 is deprecated. " <>
+            "Please use :on keyword option instead."
+          [on: on_expr]
+      end
+
     query
-    |> Join.build(qual, binding, expr, on, nil, __CALLER__)
+    |> Join.build(qual, binding, expr, opts[:on], nil, __CALLER__)
     |> elem(0)
   end
 
+  @join_opts [:on]
+
+  defp parse_join_opts([]), do: {:opts, []}
+  defp parse_join_opts(lst) when is_list(lst) do
```

Comment:
This is beautiful. My only comment is to not abbreviate variable names. So `list` instead of `lst`. :)